### PR TITLE
Improve IndexingMemoryController a bit

### PR DESCRIPTION
**Pull request by mikemccand**
I started by trying to explain the test failure in #13487, but failed so far, but then in the process found some other things to fix:
- The `IndexingMemoryController` sets indexing buffer sizes for each shard, but then doesn't always push those changes down to `IndexWriter`
- Instead, we push settings changes down to IW only during flush and refresh, but I think this just hides bugs, and for some configurations could be quite a while ... we should push changes immediately whenever settings are updated
- If you try to specify `indices.memory.index_buffer_size` in the node settings to a bytes (not %) value, you'll hit NPE on starting up the node
- Removed some dead code, added some missing units (time -> timeMS) in variable names
- We currently wait for all merges in a shard to finish before considering it inactive, but I think that's not necessary, i.e. we can safely drop the indexing buffer for an inactive shard while merges are still running.  Other shards can use that heap.

I think we should separately fix these (this PR) ... but I'll keep trying to explain the original test failure.

I did add some more `logger.debug` so maybe next failure reveals something.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java [hunk 1]
```diff
@@ -94,8 +94,11 @@
     public static final String INDEX_GC_DELETES_SETTING = "index.gc_deletes";
 
     /**
-     * Index setting to control the initial index buffer size.
-     * This setting is <b>not</b> realtime updateable.
+     * Index setting to control the initial index buffer size.  NOTE: this setting is somewhat
+     * useless, since IndexingMemoryController will take over quickly and partition the
+     * indices.memory.index_buffer_size for this node across all shards.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java]: do we want to remove it for 2.0 then?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -985,15 +985,23 @@ public void addFailedEngineListener(Engine.FailedEngineListener failedEngineList
     }
 
     public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValue shardTranslogBufferSize) {
+        Engine engine = engineUnsafe();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I think this is wrong? we should change the config even if the engine is not yet started (a null engine doesn't always mean shard closed). I think we run into issues here in the past..

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 3]
```diff
@@ -985,15 +985,23 @@ public void addFailedEngineListener(Engine.FailedEngineListener failedEngineList
...
-        if (preValue.bytes() != shardIndexingBufferSize.bytes() && engineUnsafe() != null) {
-            // its inactive, make sure we do a refresh / full IW flush in this case, since the memory
-            // changes only after a "data" change has happened to the writer
-            // the index writer lazily allocates memory and a refresh will clean it all up.
-            if (shardIndexingBufferSize == EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER && preValue != EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER) {
+        if (preValue.bytes() != shardIndexingBufferSize.bytes()) {
+            if (shardIndexingBufferSize == EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER) {
+                // it's inactive: make sure we do a refresh / full IW flush in this case, since the memory
+                // changes only after a "data" change has happened to the writer
+                // the index writer lazily allocates memory and a refresh will clean it all up.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: double checking this is OK - the previous code first did the settings update on the IW, then hit refresh... Also, refresh is optional (maybeRefreshBlocking) - what happens if there is a large buffer but no data to refresh?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 4]
```diff
@@ -1129,6 +1138,8 @@ public void onRefreshSettings(Settings settings) {
             searchService.onRefreshSettings(settings);
             indexingService.onRefreshSettings(settings);
             if (change) {
+                engine().onSettingsChanged();
+                // TODO: why force a refresh here...?
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: doesn't it follow the same logic as in updateBufferSize ? We want to have the buffer cleaned immediately if needed?

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 5]
```diff
@@ -91,7 +93,7 @@ public IndexingMemoryController(Settings settings, ThreadPool threadPool, Indice
                 indexingBuffer = maxIndexingBuffer;
             }
         } else {
-            indexingBuffer = ByteSizeValue.parseBytesSizeValue(indexingBufferSetting, null);
+            indexingBuffer = ByteSizeValue.parseBytesSizeValue(indexingBufferSetting, INDEX_BUFFER_SIZE_SETTING);
         }
         this.indexingBuffer = indexingBuffer;
         this.minShardIndexBufferSize = this.settings.getAsBytesSize("indices.memory.min_shard_index_buffer_size", new ByteSizeValue(4, ByteSizeUnit.MB));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: make this constants too?

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 6]
```diff
@@ -190,23 +194,24 @@ private int updateShardStatuses(EnumSet<ShardStatusChangeType> changes, List<Ind
                 for (IndexShard indexShard : indexService) {
 
                     if (!CAN_UPDATE_INDEX_BUFFER_STATES.contains(indexShard.state())) {
-                        // not ready to be updated yet.
+                        // not ready to be updated yet
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: pedantic :)

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 7]
```diff
@@ -170,11 +170,15 @@ public void run() {
                     indexShard.markAsInactive();
                 } catch (EngineClosedException e) {
                     // ignore
+                    logger.debug("ignore EngineClosedException while marking shard [{}][{}] as inactive", indexShard.shardId().index().name(), indexShard.shardId().id());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: move these to TRACE? they don'e really signal any issue but people may thing so...

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 8]
```diff
@@ -170,11 +170,15 @@ public void run() {
...
                 } catch (FlushNotAllowedEngineException e) {
                     // ignore
+                    logger.debug("ignore FlushNotAllowedException while marking shard [{}][{}] as inactive", indexShard.shardId().index().name(), indexShard.shardId().id());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: same trace thingy

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 9]
```diff
@@ -261,31 +266,28 @@ private int updateShardStatuses(EnumSet<ShardStatusChangeType> changes, List<Ind
...
+                    changes.add(ShardStatusChangeType.DELETED);
+                    statusShardIdIterator.remove();
+                }
             }
             return changes;
         }
 
         private void calcAndSetShardBuffers(int activeShards, String reason) {
             if (activeShards == 0) {
+                logger.debug("no active shards (reason={})", reason);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: trace? this runs very frequently..

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 10]
```diff
@@ -363,6 +364,7 @@ public void testSegments() throws Exception {
             assertThat(segments.get(1).isCompound(), equalTo(false));
 
             engine.config().setCompoundOnFlush(true);
+            engine.onSettingsChanged();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: wondering - why is this needed now?

File: core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerIT.java [hunk 11]
```diff
@@ -109,6 +110,43 @@ public void testIndexBufferSizeUpdateInactiveShard() throws InterruptedException
...
+        }
+        if (awaitBusy(() -> getIWBufferSize("test4") == 16*1024*1024) == false) {
+            fail("failed to update shard indexing buffer size for test4 index to 16 MB; got: " + getIWBufferSize("test4"));
+        }
+    }
+
+    @Test
+    public void testIndexBufferNotPercent() throws InterruptedException {
+        // #13487: Make sure you can specify non-percent sized index buffer and not hit NPE
+        createNode(Settings.builder().put(IndexingMemoryController.INDEX_BUFFER_SIZE_SETTING, "32mb").build());
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerIT.java]: do we want to validate what came out of this?

**Comment by bleskes** [comment]
Thx mike. Left comments/questions here and there.

File: core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java (hunk 1 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java]: > do we want to remove it for 2.0 then?

+1, but I'll open a separate issue.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 2 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Good catch, I'll fix it to set the config even if engine is null, and add javadocs about when engineUnsafe can return null...

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 3 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: >  the previous code first did the settings update on the IW, then hit refresh..

Good catch, I'll move the `onSettingsChange` up before the refresh.

The previous code (and still with this change) only does a refresh when the buffer is changed down to its inactive value (512 KB) ... other buffer changes still won't be "noticed" by IW until another doc or delete op is run against the shard.

>  Also, refresh is optional (maybeRefreshBlocking) - what happens if there is a large buffer but no data to refresh?

I think that's OK: it means IW is not holding onto any memory.  IW still gets the buffering change, and the next op that hits it will use that new buffer size.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 4 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: > doesn't it follow the same logic as in updateBufferSize ?

Hmm but `index_buffer_size` is not a live setting?

I think the `index.gc_deletes` and `index.version_map_size` changes should be separately handled ... I'll fix those explicitly, and remove the refresh.

I think it's sloppy to couple `refresh` to settings changes unless there really are settings changes that can only "become live" with a refresh.

Separately, `index.translog.fs.type` is a live setting, but we don't force a flush if that one is changed ... hmmm.

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java (hunk 9 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: I'd like to leave this one as debug: it should not run very frequently?  And there is already another debug line below for the more common case when there are active shards on this node ...

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 10 above)
Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: Because the test previously relied on `refresh` to make this settings change live, which I think is a poor API.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 3 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I see - so IW only holds memory to pending docs - no pre-allocated buffers? (excluding over allocation, which I presume it does when it regrows buffers)

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java (hunk 9 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: I missed this method is only called under the condition that something changed:

```
 if (!changes.isEmpty()) {
                calcAndSetShardBuffers(activeShards, "[" + changes + "]");
            }
```

++ to leaving on debug.

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 10 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: OK. So it's not strictly speaking needed, but you rather not rely on the refresh. Cool with me.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 3 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: > I see - so IW only holds memory to pending docs - no pre-allocated buffers?

That's correct: it only allocates memory on demand as docs are being indexed (and as deletes are being buffered), then frees it all when segment is written to disk.

> excluding over allocation, which I presume it does when it regrows buffers)

Most of its buffering is actually paged, so you don't have a case (mostly) where 2 contiguous huge arrays need to be in memory at once.

But, one case where we do intentionally go over-budget is when writing new segments to disk can't keep up with incoming indexing ... in this case we're will to go up to 50% over budget before IndexWriter blocks incoming indexing.

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 10 above)
Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: > So it's not strictly speaking needed, but you rather not rely on the refresh. 

It is needed, otherwise the test fails, with my changes here, because I removed the "secretly push settings changes to IW on refresh".

File: core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerIT.java (hunk 11 above)
Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerIT.java]: OK will do.

**Comment by mikemccand** [comment]
Thanks @bleskes, I folded in the feedback!

File: core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java (hunk 1 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java]: > +1, but I'll open a separate issue.

I opened #13563 for this.

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 12]
```diff
@@ -116,15 +152,19 @@ public IndexingMemoryController(Settings settings, ThreadPool threadPool, Indice
...
-        this.inactiveTime = this.settings.getAsTime("indices.memory.shard_inactive_time", TimeValue.timeValueMinutes(5));
+        this.inactiveTime = this.settings.getAsTime(SHARD_INACTIVE_TIME_SETTING, TimeValue.timeValueMinutes(5));
         // we need to have this relatively small to move a shard from inactive to active fast (enough)
-        this.interval = this.settings.getAsTime("indices.memory.interval", TimeValue.timeValueSeconds(30));
-
-        logger.debug("using index_buffer_size [{}], with min_shard_index_buffer_size [{}], max_shard_index_buffer_size [{}], shard_inactive_time [{}]", this.indexingBuffer, this.minShardIndexBufferSize, this.maxShardIndexBufferSize, this.inactiveTime);
-
+        this.interval = this.settings.getAsTime(SHARD_INACTIVE_INTERVAL_TIME_SETTING, TimeValue.timeValueSeconds(30));
+
+        logger.debug("using indexing buffer size [{}], with {} [{}], {} [{}], {} [{}], {} [{}]",
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: ++

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 10 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: OK again :) I misread "we push settings changes down to IW only during flush and refresh" to mean that Lucene does it that way. I was wrong.

File: core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerIT.java [hunk 13]
```diff
@@ -109,6 +110,45 @@ public void testIndexBufferSizeUpdateInactiveShard() throws InterruptedException
...
+        assertEquals(EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER.bytes(), getIWBufferSize("test1"));
+    }
+
+    private long getIWBufferSize(String indexName) {
+        return client().admin().indices().prepareStats(indexName).get().getTotal().getSegments().getIndexWriterMaxMemoryInBytes();
+    }
+
+    @Test
+    public void testIndexBufferSizeTwoShards() throws InterruptedException {
+        createNode(Settings.builder().put(IndexingMemoryController.SHARD_INACTIVE_TIME_SETTING, "100000h",
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerIT.java]: do we want to speed up indices.memory.interval as well?

**Comment by bleskes** [comment]
LGTM. Thx @mikemccand

File: core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerIT.java (hunk 13 above)
Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerIT.java]: > do we want to speed up indices.memory.interval as well?

I think it's not needed here: the test is just confirming that on init, the setting was parsed correctly (and not that e.g. the memory controller work up and did something important).

In fact: I don't know why the other test must set `100ms` since `createNode` does it as well ... I'll try to remove it!

Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerIT.java]: > In fact: I don't know why the other test must set 100ms since createNode does it as well ... I'll try to remove it!

OK I was wrong: `createNode` sets `indices.memory.interval` to 100ms, but that test sets `shard_inactive_time` to 100ms ... so I think all is good ... I'll push shortly.
